react-renderer/src/inspector/views/mod/LabeledInput.tsx

```diff
@@ -0,0 +1,258 @@
+    modAttr(attrname: string, attrval: Value<any>): void;
+}
+
+interface IInputProps {
+    inputType: string,
+    showValue?: string,
+    min? : string,
+    max? : string,
+    minX? : string,
+    minY? : string,
+    maxX? : string,
+    maxY? : string,
+    options? : string[]
+}
+
+// if we move to not hardcoding canvas size this needs to be changed
+// returns number corresponding to x/y coord boundary of canvas if necessary
+const toCanvas = (jsonVal: string) : string => {
+    // const cdiv = document.getElementsByClassName("Pane1")[0];
+    // todo check safety
+    // const cwidth = parseInt(window.getComputedStyle(cdiv).getPropertyValue("width"), 10),
+    //       cheight = parseInt(window.getComputedStyle(cdiv).getPropertyValue("height"), 10);
+    const cwidth = 800,
+        cheight = 700;
+    switch (jsonVal) {
+        case "CANVASL":
+            return (-(cwidth) / 2).toString();
+        case "CANVASR":
+                return (cwidth / 2).toString();
+        case "CANVASB":
+            return (-(cheight)/ 2).toString();
+        case "CANVAST":
+            return (cheight/ 2).toString();
+        default:
+            return jsonVal;
+    }
+}
+
+class LabeledInput extends React.Component<IProps> {
+    public readonly state = {
+        eValue: this.props.eValue
+    }
+    public componentDidUpdate(prevProps: IProps) {
+        if (this.props !== prevProps) {
+            this.setState({
+                eValue: this.props.eValue
+            });
+        };
+    } // todo - will subpath<x> always have x = number?
+    public updateAttr = (id: string, evalue: string | number | Color<number> | boolean | SubPath<number>[]) => {
+        // this.setState((prevState) => ({
+        //     eValue: {
+        //         ...prevState.eValue,
+        //         contents: evalue
+        //     }
+        // }))
+        
+        // this way is not safe apparently. https://stackoverflow.com/questions/43638938/updating-an-object-with-setstate-in-react
+        // however, the alternative is above and gives an error!!! 
+        // anyway, this is a possible source of a bug
+        const newstate = {
+            eValue: {
+                ...this.state.eValue,
+                contents: evalue
+            }
+        }
+        this.setState(newstate);  // will update span values - could be phased out if spans are set manually
+        this.props.modAttr(id, newstate.eValue as Value<any>);
+    }
+    public handleChange = (event: React.ChangeEvent<HTMLInputElement|HTMLSelectElement>) => {
+        const teval = (isNaN(Number(event.target.value))) ? event.target.value : +event.target.value; // convert to number if necessary
+        this.updateAttr(event.target.id, teval); // a little hacky - id is name of property
+    }
+    // when an input corresponding to a part of an attribute is modified instead of an entire attribute
+    public handleSubChange = (event: React.ChangeEvent<HTMLInputElement>) => {
```

Comment:
I have no idea what this does but thats fine I guess

Reply:
Currently it modifies inputs that map to only part of a shape attribute, e.g. how both the `opacity` and `color` input map to the `color` attribute. I plan on factoring it out.